src/main/java/com/exercise/sfe/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,234 @@
+import com.exercise.sfe.entity.GiftCertificate;
+import com.exercise.sfe.entity.Tag;
+import com.exercise.sfe.entity.dto.GiftCertificateDto;
+import com.exercise.sfe.entity.search.SearchingSettings;
+import com.exercise.sfe.exception.PersistenceException;
+import com.exercise.sfe.mapper.GiftCertificateMapper;
+import com.exercise.sfe.repository.GiftCertificateRepository;
+import com.exercise.sfe.repository.TagRepository;
+import com.exercise.sfe.service.GiftCertificateService;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@AllArgsConstructor
+@Transactional
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+  private GiftCertificateRepository giftCertificateRepository;
+  private TagRepository tagRepository;
+  private GiftCertificateMapper giftCertificateMapper;
+
+  public List<GiftCertificate> getAll() {
+    return giftCertificateRepository.getAll()
+        .stream()
+        .peek(s -> {
+          try {
+            s.setTags(tagRepository.getAllByGiftCertificateId(s.getId()));
+          } catch (Exception e) {
+            throw new PersistenceException(
+                ErrorCodes.TAG_NOT_FOUND,
+                String.format(ErrorMessages.TAG_NOT_FOUND, s.getId()));
+          }
+        })
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public List<GiftCertificate> getAllAccordingToSettings(SearchingSettings settings) {
+    if (settings == null) {
+      return getAll();
+    } else {
+      return getAll(settings);
+    }
+  }
+
+
+  @Override
+  public List<GiftCertificate> getAll(SearchingSettings settings) {
+
+    return giftCertificateRepository.getAll(settings)
+        .stream()
+        .peek(s -> {
+          try {
+            s.setTags(tagRepository.getAllByGiftCertificateId(s.getId()));
+          } catch (Exception e) {
+            throw new PersistenceException(
+                ErrorCodes.SEARCH_FILTER_WRONG_DATA,
+                String.format(ErrorMessages.SEARCH_FILTER_WRONG_DATA, s.getId()));
+          }
+        })
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public GiftCertificate getById(Long id) {
+    return giftCertificateRepository.getById(id)
+        .map(this::enrichByTags)
+        .orElseThrow(
+            () -> new PersistenceException(
+                ErrorCodes.GIFT_CERTIFICATE_NOT_FOUND,
+                String.format(ErrorMessages.GIFT_CERTIFICATE_NOT_FOUND, id))
+        );
+  }
+
+  private GiftCertificate enrichByTags(GiftCertificate giftCertificate) {
+    giftCertificate.setTags(tagRepository.getAllByGiftCertificateId(giftCertificate.getId()));
+    return giftCertificate;
+  }
+
+  @Transactional
+  @Override
+  public GiftCertificate update(GiftCertificateDto giftCertificateDto, Long id) {
```

Comment:
Can you refactor it? This method is complex and its difficult to read it. Should be short and clear